File: Algorithms/leetcode/Ex_19_Remove_Nth_Node_From_End_of_List/Soln_1_Two_Pointers.java
package Algorithms.leetcode.Ex_19_Remove_Nth_Node_From_End_of_List;

/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution_1 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) {
            return head;
        }
        ListNode fast = head;
        ListNode slow = head;
        int cnt = 0;
        // 首先先让fast指针先走n步
        while (fast != null && cnt < n) {
            fast = fast.next;
            cnt += 1;
        }
        // 然后slow和fast各自向前走，直到fast.next为空
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        if (fast == null) {
            // 说明要删除的是头节点，直接返回head.next
            return head.next;
        } else if (fast.next == null) {
            // 删除slow.next节点
            slow.next = slow.next.next;
        }
        return head;
    }
}

File: Algorithms/leetcode/Ex_0092_Reverse_Linked_List_II/Soln_1_Two_Pointers.cpp
#include <iostream>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution_1 {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (left == right) {
            return head;
        }
        if (head == nullptr || head->next == nullptr) {
            return head;
        }

        ListNode *preLeft = nullptr;
        ListNode *slow = head;
        ListNode *fast = head->next;
        int cnt = 1;

        // 首先将 slow 移动到第 left 个 node
        while (cnt < left) {
            preLeft = slow;
            slow = slow->next;
            fast = fast->next;
            cnt ++;
        }

        // 记住当前的 slow 是 第left个node
        ListNode *leftNode = slow;
        
        // 翻转 从 第left+1 到 第right 的next指针
        // 1 -> 2 -> 3 -> 4 -> 5 -> x
        // => 
        // 1 -> 2 <- 3 <- 4    5 -> x
        // ^    ^         ^    ^ 
        // preLeft             postRight
        //      leftNode  rightNode
        ListNode *nextFast;
        while (cnt < right) {
            nextFast = fast->next;
            fast->next = slow;
            
            slow = fast;
            fast = nextFast;
            cnt ++;
        }
        ListNode *postRight = fast;    
        ListNode *rightNode = slow;

        // 将 2 的 next 连到 5，1 的 next 连到 4
        leftNode->next = postRight;
        if (preLeft != nullptr) {
            preLeft->next = slow;
        }

        if (left == 1) {
            // 说明从第一个 node 就开始反转了，所以新的头节点是 rightNode
            return rightNode;
        } else {
            // 头节点没有变
            return head;
        }
    }
};
